Added OMPT integration information to geopm_prof doc and removed MPI dependency. (#3574)
geopm/geopm: docs/source/geopm_prof.3.rst

-----------

The functions described here enable application feedback to the GEOPM
control algorithm for:

> * identifying regions of code
> * determining progress within regions
> * counting iterations through loops that contain inter-node synchronization
>   points in the application.

Reviewer: ```suggestion
* Identifying regions of code
* Determining progress within regions
* Counting iterations through loops that contain inter-node synchronization
  points in the application
```

A region is defined here as a demarcated section of code, typically
with distinct compute, memory, or network characteristics. Defining

> regions of code can be useful for understanding the behavior of a section

Reviewer: ```suggestion
regions of code can be useful for: understanding the behavior of a section
```

of code, tuning control parameters to optimize the behavior of a region
independently of the rest of the code, to alleviate load imbalance in the
application (under the assumption that the region is bulk synchronous), etc.

> Often, a set of regions will be marked up within an outer loop, executed a
> fixed number of times per *epoch*. This can be leveraged to learn and optimize

Reviewer: ```suggestion
Often, a set of regions will be marked up within an *epoch* (i.e. outer loop), and
will execute a fixed number of times per *epoch*. This can be leveraged to learn and optimize
```

  progress.  This method signals the completion of one work unit out
  of the total passed to ``geopm_tprof_init()``.

> OMPT Integration

Reviewer: ```suggestion
OMP Integration
```

----------------

Integrating with OMPT provides automation for region identification, entry,

> and exit. With OMPT support, a geopm report can provide per-region metrics

Reviewer: ```suggestion
and exit. With OMPT support, a GEOPM report can provide per-region metrics
```

> Integrating with OMPT provides automation for region identification, entry,
> and exit. With OMPT support, a geopm report can provide per-region metrics

Reviewer: ```suggestion
GEOPM is able to track OMP offload calls from OMP enabled applications
via the OMPT interface.  OMPT integration in GEOPM provides automation
for region identification, entry, and exit. With OMPT support, a GEOPM 
report can provide per-region metrics
```

(such as region runtime, CPU/GPU frequency, power/energy consumption, etc,
without the need to mark up the application. However, for each region where
region progress is desired, the application must be explicitly marked up with

> ``geopm_tprof_post()``.

Reviewer: Please add a pointer to this test as an example of how to do that:
https://github.com/geopm/geopm/blob/dev/integration/test/test_progress.cpp#L28

  progress.  This method signals the completion of one work unit out
  of the total passed to ``geopm_tprof_init()``.

OMP Integration
----------------

GEOPM is able to track OMP offload calls from OMP enabled applications
via the OMPT interface.  OMPT integration in GEOPM provides automation
for region identification, entry, and exit. With OMPT support, a GEOPM 
report can provide per-region metrics
(such as region runtime, CPU/GPU frequency, power/energy consumption, etc,
without the need to mark up the application. However, for each region where
region progress is desired, the application must be explicitly marked up with
``geopm_tprof_post()``.

Examples of progress markup with/without OMP are provided here: 

> `<geopm/integration/test/test_progress.cpp><https://github.com/geopm/geopm/blob/dev/integration/test/test_progress.cpp>`_

Reviewer 2: ```suggestion
`geopm/integration/test/test_progress.cpp <https://github.com/geopm/geopm/blob/dev/integration/test/test_progress.cpp>`_
```

  progress.  This method signals the completion of one work unit out
  of the total passed to ``geopm_tprof_init()``.

OMPT Integration
----------------

> Integrating with OMPT provides automation for region identification, entry,

Author: Should we add details or a link on how to integrate with ompt?

Reviewer: It's mostly automatic.  If a code uses OMP, and OMPT support is enabled in the Runtime (today, it is by default), then OMPT regions will appear in the report.  There's nothing else for the user to do unless they want progress reporting, which you discuss below.
